src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+
+private:
+    CHIP_ERROR ReadTrustedTimeSource(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadDefaultNtp(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadTimeZone(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadDSTOffset(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+};
+
+TimeSynchronizationAttrAccess gAttrAccess;
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadTrustedTimeSource(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto tts       = TimeSynchronizationServer::Instance().GetTrustedTimeSource();
+    if (!tts.IsNull())
+    {
+        err = aEncoder.Encode(tts.Value());
+    }
+    else
+    {
+        err = aEncoder.EncodeNull();
+    }
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadDefaultNtp(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto dntp      = TimeSynchronizationServer::Instance().GetDefaultNtp();
+    if (!dntp.IsNull())
+    {
+        const char * buf = reinterpret_cast<const char *>(dntp.Value().data());
+        err              = aEncoder.Encode(chip::CharSpan(buf, strnlen(buf, dntp.Value().size())));
+    }
+    else
+    {
+        err = aEncoder.EncodeNull();
+    }
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadTimeZone(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
+        DataModel::List<TimeSynchronization::Structs::TimeZoneStruct::Type> tzList =
+            TimeSynchronizationServer::Instance().GetTimeZone();
+        for (auto it = tzList.begin(); it != tzList.end(); ++it)
+        {
+            ReturnErrorOnFailure(encoder.Encode(TimeZoneCodec(*it)));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadDSTOffset(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = aEncoder.EncodeList([](const auto & encoder) -> CHIP_ERROR {
+        DataModel::List<TimeSynchronization::Structs::DSTOffsetStruct::Type> dst =
+            TimeSynchronizationServer::Instance().GetDSTOffset();
+        for (auto it = dst.begin(); it != dst.end(); ++it)
+        {
+            ReturnErrorOnFailure(encoder.Encode(*it));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (aPath.mClusterId != TimeSynchronization::Id)
+    {
+        return CHIP_ERROR_INVALID_PATH_LIST;
+    }
+
+    switch (aPath.mAttributeId)
+    {
+    case UTCTime::Id: {
+        DataModel::Nullable<uint64_t> currentUtcTime;
+        UTCTime::Get(aPath.mEndpointId, currentUtcTime);
```

Comment:
I don't understand this logic, and it probably needs an explanation for what is going on here.  What this is doing is:

1.  Implementing an AttributeAccessInterface for the attribute.
2. In that, doing a read of that attribute from the normal attribute store.
3. If the attribute store has a non-null value, reading the new time and storing it to the attribute store (which generates subscription reports, which is a spec violation).

Why are we touching the attribute store at all here?

Reply:
Yes it is a bit weird right now. I am doing this because there is no tick for UTCTime constantly updating it.
As for '3' this attribute doesn't generate reporting but everything else you said is right. 
I am open for an elegant solution here.

Reply:
> I am doing this because there is no tick for UTCTime constantly updating it.

Yes, but why would we want to be "constantly updating it"?  Per spec, the value should just be read when someone tries to read it, and there should be no subscription reports to it (see C quality).  So there's absolutely no reason to store this in the attribute store....

> As for '3' this attribute doesn't generate reporting

It will if you change the value in the attribute store.

>  I am open for an elegant solution here.

I don't see what the problem that needs solving is.  Can you explain that part, please?